tests: Bluetooth: Mesh: remove key ID distribution in mesh bsim

alxelax opened the pull request:
PR:
- removes key ID distribution functionality since ITS was instantiated and no more ID conflicts
- removes ___weak_ attribute that required for key ID distribution functionality

jhedberg commented on subsys/bluetooth/mesh/crypto_psa.c:
```diff
@@ -28,10 +28,9 @@ LOG_MODULE_REGISTER(bt_mesh_crypto_psa);
 		2 * CONFIG_BT_MESH_APP_KEY_COUNT + 2 + BT_MESH_CDB_KEY_ID_RANGE_SIZE)
-#define BT_MESH_PSA_KEY_ID_USER_MIN (PSA_KEY_ID_USER_MIN + \
-		CONFIG_BT_MESH_PSA_KEY_ID_USER_MIN_OFFSET)
+#define BT_MESH_PSA_KEY_ID_MIN (0x1FFF4000ul)
```
Would it make sense to have this as a Kconfig option?

jhedberg replied:
Btw, where did you get this value from? Do you have some specification document (if, so please provide a reference), or was this just arbitrary?

jhedberg replied:
Makes perfect sense to coordinate this within Zephyr, but in that case I think it'd be good to have a document in the tree that defines the various allocations.

jhedberg replied:
I guess you could also have a system-wide common header file with the allocations and have each user include that header (they'll just use their specific define from it).

alxelax replied:
@jhedberg FYI feature freeze for 4.1.0 - 2025/02/14
After 4.1.0 all related changes should go over API deprecation procedure

omkar3141 replied:
This is internal define for now. Collecting all such things together and putting them in single header can be done later. But if we do it now, and later change our mind, would result into deprecation procedure. I suggest step-by-step approach. First we could do this, and then after wider discussion, move all defines to seperate headers. 

We want to avoid products using different ranges for the same subsystem one PSA support for mesh becomes non-experimental (which is also planned for 4.1).

jhedberg replied:
> But if we do it now, and later change our mind, would result into deprecation procedure

That's not quite true. A new API can't immediately jump to Stable, rather it'd have to start as experimental as per https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html

Anyway, the important thing IMO is that the existing coordination is explicit somehow. This can just as well be a document somewhere in the tree that collects the various ranges used by various subsystems, and you could have a reference to it as a code comment above the place where you have the "magic value" in the code.

tomi-font replied:
I'm the one who suggested this change.
[As I've](https://github.com/zephyrproject-rtos/zephyr/pull/76222#:~:text=Establishing%20a%20way) [mentioned before](https://github.com/zephyrproject-rtos/zephyr/pull/81235#issue-2649760239) this is something I have been wanting to add to Zephyr, just haven't quite gotten to yet. I hope to make this happen for 4.2.
The goal here was to avoid committing to a scheme that might then be removed in case the solution we go for e.g. doesn't allow configurability of the IDs by the end users.

So to play it safe for 4.1 I thought it best to stick to a non-configurable value for now. It's simple to make it configurable later, but the other way around is more tedious if some users start using that and change the ID value to something else (which may also clash).
(Also, the previous default base ID used by BT Mesh was `PSA_KEY_ID_USER_MIN`, the first ID, which is not great as it's the value everyone goes for by default.)

I am planning to discuss this with relevant stakeholders (including the security WG), which is why I didn't want to make this approach (based on static defines) too official in case it's not the one we eventually take.
I totally agree that we want to coordinate within Zephyr in an explicit way, and not have e.g. defines spread over the codebase without central reference point. This was just meant as a quick preventive fix for 4.1 to maximize forward compatibility in the future. (If sticking to this approach then we'd rather have a single header file which holds the different allocated ranges.)

This value is obtained simply by taking the middle of the [PSA key ID user range](https://arm-software.github.io/psa-api/crypto/1.2/api/keys/ids.html#c.PSA_KEY_ID_USER_MIN).

omkar3141 reviewed (dismissed):
Looks good to me as discussed.

alxelax commented on tests/bsim/bluetooth/mesh/src/distribute_keyid.c:
```diff
@@ -24,9 +24,8 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #define BT_MESH_KEY_ID_RANGE_SIZE (2 * CONFIG_BT_MESH_SUBNET_COUNT + \
 		2 * CONFIG_BT_MESH_APP_KEY_COUNT + 1 + BT_MESH_CDB_KEY_ID_RANGE_SIZE)
-#define BT_MESH_PSA_KEY_ID_USER_MIN (PSA_KEY_ID_USER_MIN + \
-		CONFIG_BT_MESH_PSA_KEY_ID_USER_MIN_OFFSET)
-#define BT_MESH_TEST_PSA_KEY_ID_USER_MIN (BT_MESH_PSA_KEY_ID_USER_MIN + \
+#define BT_MESH_PSA_KEY_ID_MIN (0x1FFF4000ul)
```
This is done for bsim tests purposes. I wouldn't like to move this definition out of C file in mesh core just to make it available in bsim tests.

tomi-font commented:
Superseded by https://github.com/zephyrproject-rtos/zephyr/pull/85581.

alxelax commented:
> Superseded by #85581.

Hi @tomi-font, mostly yes. However, mesh requires some cleaning up due to your changes. I'll wait until your PR is merged a then rebase and continue cleaning changes here.
@jhedberg, I will close discussion topic that you started as solved since this is not relevant anymore.

alxelax committed:
tests: Bluetooth: Mesh: remove key id distribution

Commit removes centralized PSA key ID distribution
in mesh bsim tests since secure storage is
instantiated on every simulated device.

Signed-off-by: Aleksandr Khromykh <aleksandr.khromykh@nordicsemi.no>

alxelax committed:
Bluetooth: Mesh: remove weak attribute and rename functions

Commit removes weak attribute and renames some functions
in crypto_psa.c since there is no centralized distribution
of the PSA key ID in bsim tests and no necessity to
reimplement native mesh approach.

Signed-off-by: Aleksandr Khromykh <aleksandr.khromykh@nordicsemi.no>

PavelVPV commented on subsys/bluetooth/mesh/crypto_psa.c:
```diff
@@ -353,7 +353,7 @@ int bt_mesh_dhkey_gen(const uint8_t *pub_key, const uint8_t *priv_key, uint8_t *
 }
 
-__weak psa_key_id_t bt_mesh_user_keyid_alloc(void)
+static psa_key_id_t bt_mesh_psa_keyid_alloc(void)
```
Would it make more sense to not use `bt_mesh_` prefix for internal functions to make it easier to read the code? Because if a function name starts with `bt_mesh_` prefix it sounds like a global function.

alxelax replied:
yep, probably, you're right. Let's do as the next improvement iteration.

tomi-font replied:
Agreed. And might be good to mention this change in the migration guide?

alxelax replied:
I do not think so. These are internal functions. The only reason they weren't static this is controlling ID value form bsim tests.
